### ISSUE TITLE
[Q] What are the necessary permissions?

### ISSUE THREAD

**Issue by BernhardWebstudio** [matteosister/GitElephant#127]
What permissions have to be given to

- `.git` -directory
- `git` cmd
- repository directory

so that everything can work as expected?

**Comment by matteosister** [comment]
the permission that you would give to the user that uses git from the command line should be given to the server user. What problem do you have?

**Comment by BernhardWebstudio** [comment]
Thank you for your response (and the program as a whole). I will just sketch the developement and my findings for future interests:

It started with `The path provided (/var/ftp/restOfPath) is not a valid git repository path`. This could be resolved by adding rwx permissons for www-data (the server-user) to the path.

Next up was `Not a git repository (or any of the parent directories): .git` which could be resolved by adding rwx permissions to the .git directory. 

`unable to access '.git/config': Permission denied ` was resolved, once again, by setting permissions to rwx. 

`Uncaught PHP Exception RuntimeException: "Exit code: 1 while executing: "/usr/bin/git pull '--rebase'" with reason: anyFile: Permission denied` was the next upcoming which had to be resolved by adding x permission to all files in the repository. 

So, in essence, to answer my own question, the whole repository including .git needs rwx-permissions. I am not exactly sure why every file needs the e_x_ecute permission, but it is like it is.

Currently, I am struggling with `Uncaught PHP Exception RuntimeException: "Exit code: 1 while executing: "/usr/bin/git pull '--rebase'" with reason: Could not create directory '/var/www/.ssh'. Host key verification failed. fatal: Could not read from remote repository.  Please make sure you have the correct access rights and the repository exists.`. I suspect some changes to the _id_rsa_ file are necessary and will update this post as soon as I found out.

**Comment by BernhardWebstudio** [comment]
Okay, the last follow-up as I finally managed to get everything working as intended:
To fix the ssh error, one has to create new ssh keys for the apache user like so: `sudo -u apache ssh-keygen -t rsa`. These will probably be saved to the home directory of the apache user, respectively e.g. `/var/www/.ssh/`. To enable the ssh-key, the public part has to be loaded at the remote-side, e.g. GitHub, [as you do](https://help.github.com/articles/adding-a-new-ssh-key-to-your-github-account/). Afterwards, the whole ssh-directory had to be added to the global ssh_config file (which has to be readable by the apache user) under Host with key [IdentityFile](https://www.cyberciti.biz/faq/force-ssh-client-to-use-given-private-key-identity-file/). 
In the end, I also got some ACL which prevented some commands, but that's probably not relevant for anyone else.
